File: 1918.py
import sys
input = sys.stdin.readline

infix = input().rstrip()

def icp(op):
    if op == '(':
        return 0
    if op == '*' or op == '/':
        return 2
    if op == '+' or op == '-':
        return 3
    if op == '#':
        return 8
    return

def isp(op):
    if op == '(':
        return 8
    if op == '*' or op == '/':
        return 2
    if op == '+' or op == '-':
        return 3
    return

postfix = ''
stack = []
for c in infix:
    if c == ')':
        while stack and stack[-1] != '(':
            postfix += stack.pop()
        stack.pop()
    elif c == '+' or c == '-' or c == '*' or c == '/':
        while stack and isp(stack[-1]) <= icp(c):
            postfix += stack.pop()
        stack.append(c)
    elif c == '(':
        stack.append(c)
    else:
        postfix += c

while stack:
    postfix += stack.pop()

print(postfix)

File: 1918.cpp
#include <bits/stdc++.h>

using namespace std;

int icp(char op) {
    switch (op)
    {
    case '(':
        return 0;
        break;
    case '*':
    case '/':
        return 1;
        break;
    case '+':
    case '-':
        return 2;
        break;
    default:
        break;
    }
}

int isp(char op) {
    switch (op)
    {
    case '(':
        return 3;
        break;
    case '*':
    case '/':
        return 1;
        break;
    case '+':
    case '-':
        return 2;
        break;
    default:
        break;
    }
}

int main() {
	ios::sync_with_stdio(0);
	cin.tie(0); cout.tie(0);

	string s;
	cin >> s;

	stack<char> stc;

	for (char c : s) {
		if (c == ')') {
			for (; !stc.empty() and stc.top() != '('; stc.pop())
				cout << stc.top();
			stc.pop();
		}

		else if ('A' <= c and c <= 'Z') {
			cout << c;
		}

		else {
			for (; !stc.empty() and isp(stc.top()) <= icp(c); stc.pop())
				cout << stc.top();
			stc.push(c);
		}
	}

	while (!stc.empty()) {
		cout << stc.top();
		stc.pop();
	}
}
